runtime/permissions.rs

```diff
@@ -1196,55 +1194,87 @@ fn permission_prompt(message: &str) -> bool {
   };
 
   #[cfg(unix)]
-  fn clear_stdin() {
+  fn confirm_prompt() -> bool {
+    // For security reasons we must consume everything in stdin so that previously
+    // buffered data cannot affect the prompt.
     let r = unsafe { libc::tcflush(0, libc::TCIFLUSH) };
     assert_eq!(r, 0);
+
+    loop {
+      let mut input = String::new();
+      let stdin = std::io::stdin();
+      let result = stdin.read_line(&mut input);
+      if result.is_err() {
+        return false;
+      };
+      let ch = match input.chars().next() {
+        None => return false,
+        Some(v) => v,
+      };
+      match ch.to_ascii_lowercase() {
+        'y' => return true,
+        'n' => return false,
+        _ => {
+          // If we don't get a recognized option try again.
+          let msg_again = format!("Unrecognized option '{}' {}", ch, opts);
+          eprint!("{}", colors::bold(&msg_again));
+        }
+      };
+    }
   }
 
   #[cfg(not(unix))]
-  fn clear_stdin() {
+  fn confirm_prompt() -> bool {
+    use winapi::um::consoleapi::ReadConsoleInputW;
+    use winapi::um::wincon::FlushConsoleInputBuffer;
+    use winapi::um::wincontypes::KEY_EVENT;
+
     unsafe {
       let stdin = winapi::um::processenv::GetStdHandle(
         winapi::um::winbase::STD_INPUT_HANDLE,
       );
-      let flags =
-        winapi::um::winbase::PURGE_TXCLEAR | winapi::um::winbase::PURGE_RXCLEAR;
-      winapi::um::commapi::PurgeComm(stdin, flags);
+
+      // For security reasons we must consume everything in stdin so that
+      // previously buffered data cannot affect the prompt.
+      assert_eq!(FlushConsoleInputBuffer(stdin), 1);
+
+      // At this point, we can't read anything from std::io::std or even
+      // winapi's `ReadConsole` because it will still contain characters
+      // flushed from the input buffer. There doesn't seem to be a way to
+      // flush this other buffer or read from ReadConsole without potentially
+      // blocking, so instead we deal with the lower level input API, and
+      // listen for key events.
+      loop {
+        let mut events_read = 0;
+        let mut input_record = std::mem::zeroed();
+        assert_eq!(
+          ReadConsoleInputW(stdin, &mut input_record, 1, &mut events_read),
+          1
```

Comment:
Nit: since this function returns a `BOOL` value, it would be more readable to compare to `winapi::shared::minwindef::TRUE`.